garrysmod/gamemodes/terrortown/gamemode/vgui/sb_main.lua

```diff
@@ -132,35 +197,127 @@ function PANEL:Init()
+      return plya:Deaths() - plyb:Deaths()
+   end
+   self.sort_table["score"] = function  ( plya, plyb )
+      return plya:Frags() - plyb:Frags()
+   end
+   self.sort_table["role"] = function  ( plya, plyb )
+      local comp = (plya:GetRole() or 0) - (plyb:GetRole() or 0)
+      -- Reverse on purpose; 
+      --    otherwise the default ascending order puts boring innocents first
+      comp = 0 - comp
+      return comp
+   end
+   self.sort_table["karma"] = function  ( plya, plyb )
+      return (plya:GetBaseKarma() or 0) - (plyb:GetBaseKarma() or 0)
+   end
+   self.sort_table["color"] = function  ( plya, plyb )
+      -- Sort by HSL Hue value; to make a rainbow-ish sort
+      return Hue_For_Player(plya) - Hue_For_Player(plyb)
+   end
+
    -- the various score column headers
    self.cols = {}
-   self:AddColumn( GetTranslation("sb_ping") )
-   self:AddColumn( GetTranslation("sb_deaths") )
-   self:AddColumn( GetTranslation("sb_score") )
+   self:AddColumn( GetTranslation("sb_ping"), nil, nil,         "ping" )
+   self:AddColumn( GetTranslation("sb_deaths"), nil, nil,       "deaths" )
+   self:AddColumn( GetTranslation("sb_score"), nil, nil,        "score" )
 
    if KARMA.IsEnabled() then
-      self:AddColumn( GetTranslation("sb_karma") )
+      self:AddColumn( GetTranslation("sb_karma"), nil, nil,     "karma" )
    end
 
-   -- Let hooks add their column headers (via AddColumn())
+   self.sort_headers = {}
+   -- Reuse some translations
+   self:AddFakeColumn( GetTranslation("sb_sortby"), nil, nil,       nil ) -- "Sort by:"
+   self:AddFakeColumn( GetTranslation("equip_spec_name"), nil, nil, "name" )
+   self:AddFakeColumn( GetTranslation("col_role"), nil, nil,        "role" )
+  
+   -- Let hooks add their column headers (via AddColumn() or AddFakeColumn())
    hook.Call( "TTTScoreboardColumns", nil, self )
-
+  
    self:UpdateScoreboard()
    self:StartUpdateTimer()
 end
 
--- For headings only the label parameter is relevant, func is included for
+local function sort_header_handler(self_, lbl)
+   return function()
+      surface.PlaySound("ui/buttonclick.wav")
+
+      local sorting = GetConVar("ttt_scoreboard_sorting")
+      local ascending = GetConVar("ttt_scoreboard_ascending")
+
+      if lbl.HeadingIdentifier == sorting:GetString() then
+         ascending:SetBool(not ascending:GetBool())
+      else
+         sorting:SetString( lbl.HeadingIdentifier )
+         ascending:SetBool(true)
+      end
+
+      for _, scoregroup in pairs(self_.ply_groups) do
+         scoregroup:UpdateSortCache()
+         scoregroup:InvalidateLayout()
+      end
+
+      self_:ApplySchemeSettings()
+   end
+end
+
+-- For headings only the label parameter is relevant, second param is included for
 -- parity with sb_row
-function PANEL:AddColumn( label, func, width )
-   local lbl = vgui.Create( "DLabel", self )
+local function column_label_work(self_, table_to_add, label, _, width, sort_identifier, sort_func )

```

Comment:
The `_` is pointless here, it's a local function with no equivalent in sb_row, so there's no parity to keep.